src/lib.rs

```diff
@@ -0,0 +1,780 @@
+    pub fn reg_fixed_def(vreg: VReg, preg: PReg) -> Self {
+        Operand::new(
+            vreg,
+            OperandPolicy::FixedReg(preg),
+            OperandKind::Def,
+            OperandPos::After,
+        )
+    }
+
+    #[inline(always)]
+    pub fn vreg(self) -> VReg {
+        let vreg_idx = ((self.bits as usize) & VReg::MAX) as usize;
+        VReg::new(vreg_idx, self.class())
+    }
+
+    #[inline(always)]
+    pub fn class(self) -> RegClass {
+        let class_field = (self.bits >> 25) & 1;
+        match class_field {
+            0 => RegClass::Int,
+            1 => RegClass::Float,
+            _ => unreachable!(),
+        }
+    }
+
+    #[inline(always)]
+    pub fn kind(self) -> OperandKind {
+        let kind_field = (self.bits >> 28) & 1;
+        match kind_field {
+            0 => OperandKind::Def,
+            1 => OperandKind::Use,
+            _ => unreachable!(),
+        }
+    }
+
+    #[inline(always)]
+    pub fn pos(self) -> OperandPos {
+        let pos_field = (self.bits >> 29) & 3;
+        match pos_field {
+            0 => OperandPos::Before,
+            1 => OperandPos::After,
+            2 => OperandPos::Both,
+            _ => unreachable!(),
+        }
+    }
+
+    #[inline(always)]
+    pub fn policy(self) -> OperandPolicy {
+        let policy_field = (self.bits >> 26) & 3;
+        let preg_field = ((self.bits >> 20) as usize) & PReg::MAX;
+        match policy_field {
+            0 => OperandPolicy::Any,
+            1 => OperandPolicy::Reg,
+            2 => OperandPolicy::FixedReg(PReg::new(preg_field, self.class())),
+            3 => OperandPolicy::Reuse(preg_field),
+            _ => unreachable!(),
+        }
+    }
+
+    #[inline(always)]
+    pub fn bits(self) -> u32 {
+        self.bits
+    }
+
+    #[inline(always)]
+    pub fn from_bits(bits: u32) -> Self {
+        Operand { bits }
+    }
+}
+
+impl std::fmt::Debug for Operand {
+    fn fmt(&self, f: &mut std::fmt::Formatter) -> std::fmt::Result {
+        write!(
+            f,
+            "Operand(vreg = {:?}, class = {:?}, kind = {:?}, pos = {:?}, policy = {:?})",
+            self.vreg().vreg(),
+            self.class(),
+            self.kind(),
+            self.pos(),
+            self.policy()
+        )
+    }
+}
+
+impl std::fmt::Display for Operand {
+    fn fmt(&self, f: &mut std::fmt::Formatter) -> std::fmt::Result {
+        write!(
+            f,
+            "{:?}@{:?}: {} {}",
+            self.kind(),
+            self.pos(),
+            self.vreg(),
+            self.policy()
+        )
+    }
+}
+
+#[derive(Clone, Copy, Debug, PartialEq, Eq)]
+pub enum OperandPolicy {
+    /// Any location is fine (register or stack slot).
+    Any,
+    /// Operand must be in a register. Register is read-only for Uses.
+    Reg,
+    /// Operand must be in a fixed register.
+    FixedReg(PReg),
+    /// On defs only: reuse a use's register. Which use is given by `preg` field.
+    Reuse(usize),
```

Comment:
I've been thinking about this some more and I suspect it may actually be better to (eventually) handle this requirement with a sort of simple rematerialization. Basically if we indicate (somehow) that a vreg's one def is [arbitrary idempotent/repeatable instruction], we can do many things with this, including a load from a particular arg slot. (I have more thoughts on how this would interact with heuristics, and whether it should allow or not allow other vregs as input, but for another time...)

The alternative is to merge the `SpillSlot` and user-defined arg-area stack addressing, which is certainly possible (and I guess more conventional) but is a bit more involved.

Anyway, I'll think more about this; we can track it in an issue perhaps; but for the initial Cranelift glue I plan to just use explicit loads at the top of the function, as we do today.